src/node/node.rs

```diff
@@ -0,0 +1,951 @@
+                        .expect("Failed to commit block");
+
+                    // Previous block is sent to the validator; reset the working block
+                    self.state.working_block = WorkingBlockOption::NoWorkingBlock;
+                } else {
+                    debug!(
+                        "{}: Already committed block {:?}",
+                        self.state,
+                        BlockId::from(pbft_message.get_block().block_id.clone())
+                    );
+                }
+            }
+
+            PbftMessageType::Checkpoint => {
+                let pbft_message = protobuf::parse_from_bytes::<PbftMessage>(&msg.content)
+                    .map_err(|e| PbftError::SerializationError(e))?;
+
+                debug!(
+                    "{}: Received Checkpoint message from {:02}",
+                    self.state,
+                    self.state
+                        .get_node_id_from_bytes(pbft_message.get_info().get_signer_id())?
+                );
+
+                if self.msg_log.get_latest_checkpoint() >= pbft_message.get_info().get_seq_num() {
+                    debug!(
+                        "{}: Already at a stable checkpoint with this sequence number or past it!",
+                        self.state
+                    );
+                    return Ok(());
+                }
+
+                // Add message to the log
+                self.msg_log.add_message(pbft_message.clone());
+
+                self._handle_checkpoint(&pbft_message)?;
+            }
+
+            PbftMessageType::ViewChange => {
+                let vc_message = protobuf::parse_from_bytes::<PbftViewChange>(&msg.content)
+                    .map_err(|e| PbftError::SerializationError(e))?;
+
+                debug!(
+                    "{}: Received ViewChange message from Node {:02}",
+                    self.state,
+                    self.state
+                        .get_node_id_from_bytes(vc_message.get_info().get_signer_id())?,
+                );
+
+                self.msg_log.add_view_change(vc_message.clone());
+
+                if self.state.mode != PbftMode::ViewChange {
+                    return Ok(());
+                }
+
+                let nv_msg = self._handle_view_change(&vc_message)?;
+                let msg_bytes = nv_msg
+                    .write_to_bytes()
+                    .map_err(|e| PbftError::SerializationError(e));
+
+                match msg_bytes {
+                    Err(e) => return Err(e),
+                    Ok(bytes) => {
+                        self.state.mode = PbftMode::NewView;
+                        self._broadcast_message(&PbftMessageType::NewView, &bytes)?;
+                    }
+                }
+            }
+
+            PbftMessageType::NewView => {
+                // TODO: Here is where we should restart the multicast protocol for messages since the
+                // last stable checkpoint
```

Comment:
Is this todo still valid?

Reply:
This (and related TODOs) will be removed with an upcoming commit that deletes the `NewView` message type, since re-executing doesn't make sense for PBFT blockchain - blocks are final.